### PR TITLE
Fewer open files

### LINKED ISSUE #7167: "ResourceWarning: unclosed file" warnings in the test suite
I noticed there are many warnings that look like:

```
dask/bytes/tests/test_http.py::test_simple
  /usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <_io.FileIO [closed]>
  
  Traceback (most recent call last):
    File "/home/runner/work/dask/dask/dask/bytes/tests/test_http.py", line 50, in test_simple
      assert data == open(os.path.join(dir_server, fn), "rb").read()
  ResourceWarning: unclosed file <_io.BufferedReader name='/tmp/tmp2zbuwztd/a'>
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```

in the test suite (see [this build](https://github.com/dask/dask/runs/1827520833#step:5:168) for an example). It would be good at some point to go through and resolve these warnings.

### PR DESCRIPTION
**Pull request by jsignell**
- [ ] Closes #7167
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

File: dask/bytes/tests/test_http.py [hunk 1]
```diff
@@ -47,23 +47,25 @@ def test_simple(dir_server):
     f = open_files(root + fn)[0]
     with f as f:
         data = f.read()
-    assert data == open(os.path.join(dir_server, fn), "rb").read()
```
Comment (jsignell) [inline dask/bytes/tests/test_http.py:50]: @martindurant I was hoping that you could confirm that the changes in this file still test what the tests were intending to test.

File: dask/dataframe/io/parquet/arrow.py [hunk 2]
```diff
@@ -1783,7 +1785,7 @@ def _gather_metadata(
                     schema = metadata.schema.to_arrow_schema()
             else:
                 for piece, fn in zip(dataset.pieces, fns):
-                    md = piece.get_metadata()
+                    md = cls.collect_file_metadata(piece.path, fs, None)
```
Comment (jsignell) [inline dask/dataframe/io/parquet/arrow.py:1788]: @rjzamora it sounds like you have a better solution for this upstream?

File: dask/bytes/tests/test_http.py (hunk 1 above)
Comment (martindurant) [reply dask/bytes/tests/test_http.py:50]: Yes, all is good. I would be very surprised if these changes actually helped though - but they certainly don't hurt.

File: dask/dataframe/io/parquet/arrow.py (hunk 2 above)
Comment (rjzamora) [reply dask/dataframe/io/parquet/arrow.py:1788]: Maybe you are referring to #7295 ? (Which fixes a different problem with the `create_metadata_file` code path)

Comment (jsignell) [reply dask/dataframe/io/parquet/arrow.py:1788]: ah ok. So does this seem like a reasonable solution? The issue was that pyarrow was doing `fs.open` in `get_metadata` and then never closing.

Comment (rjzamora) [reply dask/dataframe/io/parquet/arrow.py:1788]: Ah, that makes sense.  In that case, this looks like a good solution - Thanks for addressing this!  
 
The only other change that might make sense is to take advantage of the fact that `collect_file_metadata` can handle the `set_file_path` update if we pass `fn` as the last argument.  The one complication is that we would need to pass `piece.path.replace(base + fs.sep, "")` in the case that `fn_partitioned` is True.  I don't fell strongly one way or the other, but if you use `collect_file_metadata` for the file-path update, you can remove the `set_file_path` logic below.

**Comment by GenevieveBuckley** [comment]
@jsignell a lot of these failing tests look like they might be solved with https://github.com/intake/filesystem_spec/pull/630

Right now the legacy pyarrow tests are trying to push a `ParquetDatasetPiece` object through the `stringify_path` method of `fsspec`, and then failing because they don't get a string returned.

**Comment by GenevieveBuckley** [comment]
All the other local test failures seemed to be resolved easily by merging the main branch.

I can't push to your branch, but if you merge main I think that will help.

**Comment by GenevieveBuckley** [comment]
https://github.com/intake/filesystem_spec/pull/630 is merged. After the next `fsspec` release we should re-trigger the CI and see if that's fixed things here.

**Comment by jsignell** [comment]
I took out the arrow bit for now. So this is only a partial fix, but uncontroversialal

**Review by jrbourbeau — APPROVED** [review]
Thanks @jsignell!

**Pull request merged**